### ISSUE TITLE
Swig python and distutils

### ISSUE THREAD

**Issue by eendebakpt** [swig/swig#2586]
The distutils module has been deprecated (see [PEP632](https://peps.python.org/pep-0632/)) and is planned for removal in python 3.12, but the Swig documentation still recommends distutils: https://www.swig.org/Doc4.1/Python.html#Python_nn6. What are alternative build methods that will work with python 3.12+?

**Comment by LecrisUT** [comment]
A modern python build system to use with cmake is [`scikit-build-core`](https://github.com/scikit-build/scikit-build-core).

**Comment by phetdam** [comment]
If there's still interest in this I can help update the documentation.

The changes would be more or less the following:

1. Recommend using [setuptools](https://setuptools.pypa.io/en/latest/), which has essentially replaced distutils.
2. Recommend using [CMake](https://cmake.org/cmake/help/latest/index.html), which allows Python extension module compilation to be part of larger C/C++ builds.

In particular for 2., CMake itself [has SWIG support](https://cmake.org/cmake/help/latest/module/UseSWIG.html). I recommend this for more recent version of CMake over writing one's own custom rules (unless necessary) for getting SWIG invocations to work optimally (i.e. minimal rebuild with correct dependency tracking via `-MMD`) since the alternative can be more work than one would expect.

I suppose in general custom rules warrant a discussion of the `-MMD -MF <file>` pattern as well for the dependency file.

**Comment by wsfulton** [comment]
Yes, a documentation patch will be welcome. SWIG tries to avoid getting into build systems as it is a can of worms and everyone has their own preferences. Can you convert the distutils example into something that avoids recommending CMake or anything else and keeps it all simple python like the distutil documentation example currently does atm? I'd hope that setuptools can take a simple list of .c files, module name and not much else to generate a simple extension module.

**Comment by phetdam** [comment]
> Yes, a documentation patch will be welcome. SWIG tries to avoid getting into build systems as it is a can of worms

Great! Also I agree; the C/C++/Python build ecosystems are pretty fragmented. Neither has a universal solution.

> Can you convert the distutils example

> I'd hope that setuptools can take a simple list of .c files, module name and not much else to generate a simple extension module.

Yes, definitely. In its basic form, building extension modules [in setuptools](https://setuptools.pypa.io/en/latest/userguide/ext_modules.html) is similar to how it is done in distutils. And in contrast to the newer Python build systems, setuptools is one of the few that support building C/C++ extension modules.

> I suppose in general custom rules warrant a discussion of the `-MMD -MF <file>` pattern as well for the dependency file.

Should I briefly mention the `-M` class of options? This would be helpful for anyone trying to integrate SWIG into their builds.

**Comment by wsfulton** [comment]
> > I suppose in general custom rules warrant a discussion of the `-MMD -MF <file>` pattern as well for the dependency file.
> 
> Should I briefly mention the `-M` class of options? This would be helpful for anyone trying to integrate SWIG into their builds.

Possibly if you are referring to SWIG's `-M` family of options. As this family of options is poorly documented in SWIG, a new section on them would be useful in SWIG.html, and then point to them from the Python chapter. I'm not sure how this is relevant to setuptools though, I'd have thought it would automatically build out of date files without having to go into this detail, but I can only really understand once you've written the docs! So I suggest you write what you feel is best and I'll comment properly then.

**Comment by phetdam** [comment]
> Possibly if you are referring to SWIG's `-M` family of options. 

Yes, I am referring to SWIG's `-MM`, `-MF <file>`, etc. options.

> a new section on them would be useful in SWIG.html

That sounds good. To me the `-MM`, `-MF <file>`, is SWIG's integration hook with external build systems as it provides the dependency scanning needed to know when to re-run SWIG if an `%include` changes in the `.i` file.

> I'm not sure how this is relevant to setuptools though

Yes, the `-M` options aren't specifically relevant to setuptools. But like with other build systems, if someone wanted to ensure that SWIG is re-run only when necessary, e.g. when `%include` changes occur that affect the `.i`, they would have to use `-MMD` or something similar to detect if SWIG needs to be run even if the `.i` hasn't been touched.

I think the setuptools section at most can briefly link to `-M` documentation so those that are interested can browse, but I don't think the setuptools section would be the right place for an elaborate integration example.

> So I suggest you write what you feel is best and I'll comment properly then.

Sounds like a plan :) I've already made progress on the `Python.html` changes but nothing yet on `-M` options.

**Comment by phetdam** [comment]
Ah, the `-MMD` and `-MF <file>` options *are* somewhat relevant for setuptools. Although setuptools can take `.i` files in the sources list of the [`Extension`](https://setuptools.pypa.io/en/latest/userguide/ext_modules.html#extension-api-reference), it will not make use of SWIG's dependency scanning unless you hook it up yourself. E.g.

```python
def swig_deps(depfile: Union[str | Path]) -> Iterable[str]:
    """Get a list of file dependencies from and including the SWIG depfile.

    If the depfile does not exist only the depfile name is returned. This
    ensures that setuptools will re-run the extension build rule.
    """
    deps = [str(depfile)]
    # if depfile exists, consume dependencies
    if Path(depfile).exists():
        with open(depfile) as f:
            deps += [line.strip().rstrip(" \\") for line in f.readlines()[1:]]
    return deps


# example name + depfile
name = "example"
depfile = f"{name}_wrap.c.d"
# define our C extension module. the Extension knows how to run SWIG and we
# enable dependency file generation by adding -MMD to the options
ext = Extension(
    f"_{name}",
    sources=[f"{name}.i", f"{name}.c"],
    depends=swig_deps(depfile),
    swig_opts=["-MMD", "-MF", depfile]
)
```

I realized this because I was testing the simple setuptools example in the HTML docs I will use to replace the old distutils one, and only recently noticed that the `Extension` has some built-in support for SWIG.

**Comment by phetdam** [comment]
Just to update on my progress I've made some changes to `Python.html` and `SWIG.html` to cover setuptools and the `-M` options.

I'll put up a PR for the documentation changes later for review. As a side note, the setuptools example I have replacing the original distutils one is a simplified version of what I've shown above (so no `-MMD -MF <depfile>` usage).

**Comment by phetdam** [comment]
Belated bump on this just to see if there's any interest in the Python doc changes in the linked PR.

**Comment by wsfulton** [comment]
Thanks @phetdam, #3268 is now merged to master for the swig-4.4.0 release.
